File: c/advance/Shape.c
#include <stdio.h>
#include <stddef.h>
#include <assert.h>

enum bool { TRUE=1, FALSE=0} BOOL;

typedef struct { // Shape ン⊿Τ逆
} Shape;

typedef struct { // Shape 摸Τ area() ㄧ计
  int (*area)(Shape*); // 材把计ッ环琌赣ン
} ShapeClass;

int ShapeArea(Shape *obj);

ShapeClass shapeClass = {
  .area = ShapeArea
};

int ShapeArea(Shape *obj) {
  printf("&ShapeObject=%p &shapeClass=%p\n", obj, &shapeClass);
  assert(FALSE);
}

int main() {
  Shape s;
  ShapeArea(&s);
}

File: c/advance/Shape.java
class Shape {
  double area() { return 0.0; }
  
  public static void main(String[] argv) {
    Shape s[] = { new Shape(), new Circle(3.0) };
    for (int i=0; i<s.length; i++)
      System.out.println("area()="+s[i].area());
  }
}

class Circle extends Shape {
  public double r;
  Circle(double pr) { r = pr; }
  double area() { return 3.14*r*r; }
}
 